### PR TITLE
Enable ComStock OS 3.4.0 with custom gems

### LINKED ISSUE #237: Timeseries processing doesn't get triggered for ComStock projects
**Describe the bug**
The timeseries postprocessing isn't triggered in ComStock due to misalignment of the workflow generators for commercial and residential (see #222). Basically https://github.com/NREL/buildstockbatch/blob/1cca3c82fba1d013f7f81bbad4ba05a4f6521d11/buildstockbatch/base.py#L545 is not triggered by the commercial workflow yaml which looks like this:

```yml
workflow_generator:
  type: commercial_default
  args:
    reporting_measures:
    - measure_dir_name: f8e23017-894d-4bdf-977f-37e3961e6f42 # OpenStudio Results
      arguments:
        building_summary_section: true
        annual_overview_section: true
        monthly_overview_section: true
        utility_bills_rates_section: true
        envelope_section_section: true
        space_type_breakdown_section: true
        space_type_details_section: true
        interior_lighting_section: true
        plug_loads_section: true
        exterior_light_section: true
        water_use_section: true
        hvac_load_profile: true
        zone_condition_section: true
        zone_summary_section: true
        zone_equipment_detail_section: true
        air_loops_detail_section: true
        plant_loops_detail_section: true
        outdoor_air_section: true
        cost_summary_section: true
        source_energy_section: true
        schedules_overview_section: true
    - measure_dir_name: SimulationOutputReport
    - measure_dir_name: TimeseriesCSVExport
      arguments:
        reporting_frequency: "Timestep"
        inc_output_variables: false
    - measure_dir_name: comstock_sensitivity_reports
    - measure_dir_name: qoi_report
    - measure_dir_name: la_100_qaqc
      arguments:
        run_qaqc: false
    - measure_dir_name: simulation_settings_check
      arguments:
        run_sim_settings_checks: true
```

**To Reproduce**
1. Example yaml from @mdahlhausen:

```yml
schema_version: '0.3'
buildstock_directory: /lustre/eaglefs/projects/enduse/comstock/repos/comstock
project_directory: national
output_directory: /lustre/eaglefs/projects/enduse/comstock/com_test_17_2016/
weather_files_path: /lustre/eaglefs/projects/enduse/comstock/weather/Fort_Collins_County_IDs_2016.zip

sampler:
  type: precomputed
  args:
    sample_file: /lustre/eaglefs/projects/enduse/comstock/samples/buildstock_wwr_2016_test.csv

eagle:
  account: enduse
  n_jobs: 1
  minutes_per_sim: 120
  postprocessing:
    time: 45

postprocessing:
  keep_individual_timeseries: true
  aws:
    region_name: 'us-west-2'
    s3:
      bucket: eulp
      prefix: simulation_output/regional_runs/comstock/com_test_17_2016/
    athena:
      glue_service_role: service-role/AWSGlueServiceRole-default
      database_name: enduse
      max_crawling_time: 600 # Time to wait for the crawler to complete before aborting it

baseline:
  n_buildings_represented: 30
  custom_gems: True

os_version: eulp_com_v12
os_sha: 3472e8b799
workflow_generator:
  type: commercial_default
  args:
    reporting_measures:
    - measure_dir_name: f8e23017-894d-4bdf-977f-37e3961e6f42 # OpenStudio Results
      arguments:
        building_summary_section: true
        annual_overview_section: true
        monthly_overview_section: true
        utility_bills_rates_section: true
        envelope_section_section: true
        space_type_breakdown_section: true
        space_type_details_section: true
        interior_lighting_section: true
        plug_loads_section: true
        exterior_light_section: true
        water_use_section: true
        hvac_load_profile: true
        zone_condition_section: true
        zone_summary_section: true
        zone_equipment_detail_section: true
        air_loops_detail_section: true
        plant_loops_detail_section: true
        outdoor_air_section: true
        cost_summary_section: true
        source_energy_section: true
        schedules_overview_section: true
    - measure_dir_name: SimulationOutputReport
    - measure_dir_name: TimeseriesCSVExport
      arguments:
        reporting_frequency: "Timestep"
        inc_output_variables: false
    - measure_dir_name: comstock_sensitivity_reports
    - measure_dir_name: qoi_report
    - measure_dir_name: la_100_qaqc
      arguments:
        run_qaqc: false
    - measure_dir_name: simulation_settings_check
      arguments:
        run_sim_settings_checks: true
```
2. That's about it.

**Expected behavior**
I want (but don't expect given the code) for the timeseries postprocessing to run. But it doesn't.

**Logs**

postprocessing.out
```
begin eagle_postprocessing.sh
Job ID: 6997691
Hostname: r2i7n29
UPLOADONLY: False
MEMORY: 85248
head node
r2i7n29
workers
r8i0n[11,15]
r8i0n11: Warning: Permanently added 'r8i0n11,10.148.8.126' (ECDSA) to the list of known hosts.
r8i0n15: Warning: Permanently added 'r8i0n15,10.148.8.134' (ECDSA) to the list of known hosts.
r8i0n15:               total        used        free      shared  buff/cache   available
r8i0n15: Mem:            92G        2.9G         84G        5.1G        5.2G         84G
r8i0n15: Swap:            0B          0B          0B
r8i0n11:               total        used        free      shared  buff/cache   available
r8i0n11: Mem:            92G        2.9G         85G        3.9G        4.1G         85G
r8i0n11: Swap:            0B          0B          0B
DEBUG:2021-06-15 13:01:24:buildstockbatch.base:Using OpenStudio version: eulp_com_v12 with SHA: 3472e8b799
DEBUG:2021-06-15 13:01:24:__main__:Output directory = /lustre/eaglefs/projects/enduse/comstock/com_test_17_2016
DEBUG:2021-06-15 13:01:24:__main__:Starting postprocessing
INFO:2021-06-15 13:01:28:buildstockbatch.postprocessing:Creating results_df.
INFO:2021-06-15 13:01:30:buildstockbatch.postprocessing:Writing /lustre/eaglefs/projects/enduse/comstock/com_test_17_2016/results/results_csvs/results_up00.csv.gz
INFO:2021-06-15 13:01:30:buildstockbatch.postprocessing:Uploading the parquet files to s3
INFO:2021-06-15 13:01:38:buildstockbatch.postprocessing:Upload to S3 completed. The files are uploaded to: eulp/simulation_output/regional_runs/comstock/com_test_17_2016/com_test_17_2016/
INFO:2021-06-15 13:01:38:buildstockbatch.postprocessing:Creating Athena tables using glue crawler
INFO:2021-06-15 13:01:39:buildstockbatch.postprocessing:Deleting existing tables in db enduse: ['_com_test_17_2016']. And creating new ones.
INFO:2021-06-15 13:01:51:buildstockbatch.postprocessing:Crawler started
INFO:2021-06-15 13:03:22:buildstockbatch.postprocessing:Crawler has completed running. It is STOPPING.
INFO:2021-06-15 13:03:22:buildstockbatch.postprocessing:TablesCreated: 1 TablesUpdated: 0 TablesDeleted: 0 
INFO:2021-06-15 13:06:54:buildstockbatch.postprocessing:Crawler stopped. Deleting it now.
INFO:2021-06-15 13:06:54:buildstockbatch.postprocessing:Removing results_job*.json.gz

real	5m45.335s
user	0m1.451s
sys	0m2.137s
```

**Platform (please complete the following information):**
- Simulation platform: eagle
- BuildStockBatch version, branch, or sha: 1cca3c82fba1d013f7f81bbad4ba05a4f6521d11
- resstock or comstock repo version, branch, or sha: no clue
- Local Desktop OS: n/a

**Additional context**
This is really a issue with the workflow generator schema / specification but I have no idea what the proper solution is there. Instead i'm going to role a crappy patch for this that will get us through. In general I'm not a huge fan of how timeseries postprocessing is (or really isn't) specified in the yaml but that's also a different issue for a different day.

### PR DESCRIPTION
**Pull request by asparke2**
Fixes #237

## Pull Request Description

This PR makes several modifications to work for ComStock.

First, the default commercial workflow generator was changed to mimic the residential workflow generator, where a new `timeseries_csv_export` key was added to the workflow schema in order to trigger timeseries postprocessing.  The key/value pairs under this key set the arguments to the `TimeseriesCSVExport` measure in the workflow, whereas they were previously hard-coded in the workflow generator.

```
  args:
    reporting_measures:
       - measure_dir_name: qoi_report
    - measure_dir_name: simulation_settings_check
      arguments:
        run_sim_settings_checks: true
    timeseries_csv_export:
      reporting_frequency: Timestep
      inc_output_variables: false
```

Second, the CLI commands were changed to work with OS 3.X when `custom_gems=true` in the YML file.

Third, in order to enable the use of custom gems on local docker runs, at runtime, as the first step in building the Docker image, the target BuildStock repo is mounted, and the custom gems specified by that repo (via `MyBuildStockDir/resources/Gemfile`) are installed to a volume called `buildstockbatch_custom_gems` to replace the default gems (`/var/oscli/gems `) using bundler.  The `buildstockbatch_custom_gems` volume is not deleted after the run, because this way, if a developer wants to work on the custom gems, they can do so quickly without having to re-download all the gems locally for each test run. Not coincidentally, the `MyBuildStockDir/resources/Gemfile` in the BuildStock directory is also used in the creation of custom Singularity containers used by Eagle, and is envisioned to be the source of custom gem specifications to use with BuildStockBatch for AWS runs in a future PR.

## Checklist

Not all may apply

- [x] Code changes (must work)
- [x] Tests exercising your feature/bug fix (check coverage report on CircleCI build -> Artifacts)
- [x] All other unit tests passing
- [x] Update validation for project config yaml file changes
- [x] Update existing documentation
- [x] Run a small batch run to make sure it all works (local is fine, unless an Eagle specific feature)
- [ ] Add to the changelog_dev.rst file and propose migration text in the pull request

### CODE REVIEW

**Comment by asparke2** [comment]
@nmerket I can't tell if this is failing because of something in my change or something weird with dockerhub.

**Comment by asparke2** [comment]
@nmerket I re-ran the tests and they passed, might have hit the dockerhub limit the other day.

**Review by nmerket — COMMENTED** [review]
This seems like it should be fine. Sorry it took me so long to see. Is there any way you could write a quick test for this? You'd add it somewhere around [here](https://github.com/NREL/buildstockbatch/blob/bug/enable_com_timeseries_postprocessing/buildstockbatch/workflow_generator/test_workflow_generator.py#L319)

**Comment by asparke2** [comment]
@nmerket I don't know how to test this if the CI isn't actually performing runs, any ideas? We've been using this branch for ComStock runs for about a month now, so it's working in practice.

**Comment by nmerket** [comment]
@asparke2 You can see in some of the other tests there that it's just checking the workflow output to make sure things end up in the right place. It's more lower level testing than you're thinking. I'm not going to force you to add the tests. It's code that only you guys use, so if it breaks, it's your funeral. 

Ultimately we should be doing more full stack integration testing. Maybe once we switch over to github actions so we can use docker from inside the tests properly.

**Comment by asparke2** [comment]
@nmerket I've tested this PR on Eagle and on local docker on Windows and Mac. The local docker versions both work, but are ~5-10x slower than Eagle, especially on Windows, presumably because of slow Docker drivers.

File: buildstockbatch/localdocker.py [hunk 1]
```diff
@@ -114,19 +173,22 @@ def run_building(cls, project_dir, buildstock_dir, weather_dir, docker_image, re
...
         with open(os.path.join(sim_dir, 'in.osw'), 'w') as f:
             json.dump(osw, f, indent=4)
 
         docker_client = docker.client.from_env()
-        args = [
-            'openstudio',
-            'run',
-            '-w', 'in.osw',
-        ]
+        run_cmd = 'openstudio run -w in.osw'
```
Comment (nmerket) [inline buildstockbatch/localdocker.py]: Using a list for the program arguments was an intentional choice. It allows us to more easily manipulate the arguments later without having to deal with whether we got the spaces in there correctly.

File: buildstockbatch/localdocker.py [hunk 2]
```diff
@@ -114,19 +173,22 @@ def run_building(cls, project_dir, buildstock_dir, weather_dir, docker_image, re
...
+        if cfg.get('baseline', dict()).get('custom_gems', False):
+            run_cmd = f'openstudio --bundle {mnt_custom_gem_dir}/Gemfile --bundle_path {mnt_custom_gem_dir} ' \
+                       '--bundle_without native_ext run -w in.osw --debug'
         if measures_only:
-            args.insert(2, '--measures_only')
+            run_cmd += ' --measures_only'
```
Comment (nmerket) [inline buildstockbatch/localdocker.py]: Case in point here. What if that space was missing. Very easy mistake to make.

File: buildstockbatch/localdocker.py [hunk 3]
```diff
@@ -114,19 +173,22 @@ def run_building(cls, project_dir, buildstock_dir, weather_dir, docker_image, re
...
         docker_client = docker.client.from_env()
-        args = [
-            'openstudio',
-            'run',
-            '-w', 'in.osw',
-        ]
+        run_cmd = 'openstudio run -w in.osw'
+        if cfg.get('baseline', dict()).get('custom_gems', False):
+            run_cmd = f'openstudio --bundle {mnt_custom_gem_dir}/Gemfile --bundle_path {mnt_custom_gem_dir} ' \
+                       '--bundle_without native_ext run -w in.osw --debug'
```
Comment (nmerket) [inline buildstockbatch/localdocker.py]: You'll need to make other changes to make this work, but yeah.

```suggestion
            run_cmd = [
                'openstudio',
                '--bundle', f'{mnt_custom_gem_dir}/Gemfile',
                '--bundle_path', mnt_custom_gem_dir,
                '--bundle_without native_ext',
                'run', '-w', 'in.osw', '--debug'
            ]
```

**Review by nmerket — CHANGES_REQUESTED** [review]
See some minor notes below.

You could also improve performance locally by using a [volume mount](https://docker-py.readthedocs.io/en/stable/volumes.html) into the container rather than a bind mount from a folder on the host. That creates a new volume that is only available to the containers but it doesn't have to do all the translation between the host and container, so it's more performant. I know at one point bind mounts were implemented via a network drive, so yeah, no wonder it's slow. Also it would have the benefit of not needing to have a hidden folder to keep the gems on the host, the `.custom_gems` folder. You'd just make the volume, use it, and then delete it when the batch is done.

Finally, could you add something to `changelog_dev.rst` about this? Is there anything else needed in the documentation for people to know how to use it?

**Comment by nmerket** [comment]
Oh and one more thing. It looks like our already abysmal testing coverage is even lower. I encourage you to write some simple tests to make sure your code is working as you expect. In this case you'd just write something that initializes LocalDockerBatch and make sure it does the things you want. Mocking is very useful there. I can help with that as necessary.

**Review by nmerket — COMMENTED** [review]
A couple more things and we can merge this

- [x] I'd like to run a couple small resstock runs on both localdocker and Eagle just to make sure we didn't break anything with this. I don't want everyone coming with their pitchforks.
- [x] Needs a changelog entry
- [x] Is there any other documentation that needs to be updated?

**Comment by nmerket** [comment]
Did a test run on Eagle and Local Docker. Worked Great.

**Comment by asparke2** [comment]
No documentation change necessary, already modified the example config in `commercial_default.rst`

**Comment by asparke2** [comment]
Did a ComStock local docker test run with custom gems, worked correctly.

**Comment by nmerket** [comment]
I did some ResStock runs on both Eagle and locally and it worked fine.

**Comment by nmerket** [comment]
There should definitely be some documentation in project_defn.rst about how to use the custom gems.

**Comment by nmerket** [comment]
It looks like you do have some documentation [here](https://github.com/NREL/buildstockbatch/blob/bug/enable_com_timeseries_postprocessing/docs/project_defn.rst#baseline-simulations-incl-sampling-algorithm) but it's pretty outdated at this point. Can you update it to reflect the current state of things?

**Comment by asparke2** [comment]
@nmerket I added documentation changes for custom gems and did a successful test run on Eagle.

File: buildstockbatch/test/test_postprocessing.py [hunk 4]
```diff
@@ -87,7 +87,7 @@ def test_large_parquet_combine(basic_residential_project_file):
     with patch.object(BuildStockBatchBase, 'weather_dir', None), \
             patch.object(BuildStockBatchBase, 'get_dask_client'), \
             patch.object(BuildStockBatchBase, 'results_dir', results_dir),\
-            patch.object(postprocessing, 'MAX_PARQUET_MEMORY', 1e6):  # set the max memory to just 1MB
+            patch.object(postprocessing, 'MAX_PARQUET_MEMORY', 1):  # set the max memory to just 1MB
```
Comment (nmerket) [inline buildstockbatch/test/test_postprocessing.py:90]: Good catch.

File: docs/project_defn.rst [hunk 5]
```diff
@@ -77,11 +77,14 @@ Information about baseline simulations are listed under the ``baseline`` key.
...
-  image is properly configured but that's easier said than done. The la100 branch on the nrel/docker-openstudio repo
-  is a starting place if this is required.
+- ``custom_gems``: true or false. **ONLY WORKS ON EAGLE AND LOCAL DOCKER** When true, buildstockbatch will
+  call the OpenStudio CLI commands with the  ``bundle`` and ``bundle_path`` options. These options tell the CLI
+  to load a custom set of gems rather than those included in the OpenStudio CLI. For both Eagle
+  and local Docker runs, these gems are first specified in the ``buildstock\resources\Gemfile``.
+  For Eagle, when the Singularity image is built, these gems are added to the image.
+  For local Docker, when the containers are started, the gems specified in the Gemfile are installed into a Docker
+  volume on the local computer. This volume is mounted by each container as models are run, so each run
+  uses the custom gems.
```
Comment (nmerket) [inline docs/project_defn.rst:87]: Thanks for updating the docs.

**Review by nmerket — APPROVED** [review]
Thanks for doing all the testing and everything.

**Pull request merged**